lib/webdriver_proxy.ts

```diff
@@ -0,0 +1,76 @@
+import * as http from 'http';
+import * as url from 'url';
+
+import {parseWebDriverCommand, WebDriverCommand} from './webdriver_commands';
+
+/**
+ * A proxy that understands WebDriver commands. Users can add middleware (similar to middleware in
+ * express) that will be called before
+ * forwarding the request to WebDriver or forwarding the response to the client.
+ */
+export class WebDriverProxy {
+  barriers: WebDriverBarrier[];
+  seleniumAddress: string;
+
+  constructor(seleniumAddress: string) {
+    this.barriers = [];
+    this.seleniumAddress = seleniumAddress;
+  }
+
+  addBarrier(barrier: WebDriverBarrier) {
+    this.barriers.push(barrier);
+  }
+
+  requestListener(originalRequest: http.IncomingMessage, response: http.ServerResponse) {
+    let command = parseWebDriverCommand(originalRequest.url, originalRequest.method);
+
+    let replyWithError = (err) => {
+      response.writeHead(500);
+      response.write(err.toString());
+      response.end();
+    };
+
+    // TODO: What happens when barriers error? return a client error?
+    let barrierPromises = this.barriers.map((b) => b.onCommand(command));
```

Comment:
This could result in multiple barriers executing at once, which might confuse webdriver if for example two barriers are both calling `executeScript`. I think we should gate them to running one at a time.